module/zfs/zio.c

```diff
@@ -3474,6 +3476,25 @@ zio_vdev_io_done(zio_t *zio)
 	return (ZIO_PIPELINE_CONTINUE);
 }
 
+void
+zio_change_priority(zio_t *pio, zio_priority_t priority)
+{
+	zio_t *cio, *cio_next;
+	zio_link_t *zl = NULL;
+
+	if (pio->io_vd != NULL && pio->io_vd->vdev_ops->vdev_op_leaf &&
+	    (pio->io_type == ZIO_TYPE_READ || pio->io_type == ZIO_TYPE_WRITE)) {
+		vdev_queue_change_io_priority(pio, priority);
+	}
+
+	mutex_enter(&pio->io_lock);
```

Comment:
I'm assuming you're grabbing the locks to prevent any new children while you walk them all. Is that the idea here? Is it really required?

Reply:
Yes it is. I hit a NULL pointer dereference when running the code without the lock.

Reply:
Can you add an ASSERT(MUTEX_HELD(pio->io_lock)) in `zio_walk_children`?

Reply:
Will do